src/stacked_borrows.rs

```diff
@@ -467,6 +469,51 @@ impl<'tcx> Stack {
 
         Ok(())
     }
+
+    /// Report a descriptive error when `new` could not be granted from `derived_from`.
+    fn grant_error(
+        &self,
+        derived_from: SbTag,
+        new: Item,
+        alloc_id: AllocId,
+        alloc_range: AllocRange,
+    ) -> InterpError<'static> {
+        let action = format!(
+            "trying to reborrow for {:?} over {:?}[{:#x}..{:#x}]",
+            new.perm,
+            alloc_id,
+            alloc_range.start.bytes(),
+            alloc_range.end().bytes()
+        );
+        // Find all permissions that this tag grants. When we are not doing raw pointer tagging,
+        // there may be many untagged items in the borrow stack, and we want to report the most
+        // permissions that the tag grants.
+        let mut available_permissions = self
+            .borrows
+            .iter()
+            .filter_map(|item| (item.tag == derived_from).then(|| item.perm))
+            .filter(|perm| *perm != Permission::Disabled) // Dont't want to say "parent tag grants Disabled"
+            .collect::<Vec<_>>();
+        available_permissions.sort_by_key(|perm| {
+            match perm {
+                Permission::Unique => 0,
+                Permission::SharedReadWrite => 1,
+                Permission::SharedReadOnly => 2,
+                Permission::Disabled => unreachable!(),
+            }
+        });
+        if let Some(max_granted) = available_permissions.get(0) {
+            err_sb_ub(format!(
+                "{}, but parent tag {:?} grants insufficient permission: {:?}",
```

Comment:
"insufficient permission: Unique" and "insufficient permission: SharedReadWrite" seems impossible, doesn't it?

Reply:
Sure does.